src/main/java/com/amplitude/HttpTransport.java

```diff
@@ -109,100 +111,30 @@ private void onEventsError(List<Event> events, Response response) {
-
-                  if (eventIndicesToRemove.length > 0) {
-                    List<Event> eventsToDrop = new ArrayList<>();
-                    int numEventsRemoved = 0;
-                    for (int i = eventIndicesToRemove.length - 1; i >= 0; i--) {
-                      int index = eventIndicesToRemove[i];
-                      if (index < eventCount) {
-                        eventsToDrop.add(eventsBuffer.remove(index));
-                        numEventsRemoved += 1;
-                      }
-                    }
-                    triggerEventCallbacks(eventsToDrop, retryResult.statusCode, "Invalid events.");
-                    eventCount -= numEventsRemoved;
-                    eventsInRetry.addAndGet(-numEventsRemoved);
-                  }
-                  if (!shouldRetry || eventCount < 1) {
-                    break;
-                  }
-                  if (shouldReduceEventCount && !isLastTry) {
-                    triggerEventCallbacks(
-                        eventsBuffer.subList(eventCount / 2, eventCount),
-                        retryResult.statusCode,
-                        "Event dropped for retry");
-                    eventCount /= 2;
-                  }
-                  if (isLastTry) {
-                    triggerEventCallbacks(
-                        eventsBuffer.subList(0, eventCount),
-                        retryResult.statusCode,
-                        "Event retries exhausted.");
-                  }
-                  if (eventCount == 0) {
-                    break;
-                  }
-                } catch (InterruptedException | AmplitudeInvalidAPIKeyException e) {
-                  // The retry logic should only be executed after the API key checking passed.
-                  // This catch AmplitudeInvalidAPIKeyException is just for handling
-                  // retryEventsOnce in thread.
-                  Thread.currentThread().interrupt();
-                }
-              }
-              eventsInRetry.addAndGet(-eventCount);
-              if (recordThrottledId) {
-                synchronized (throttleLock) {
-                  if (throttledUserId.containsKey(userId)) {
-                    throttledUserId.remove(userId);
-                  }
-                  if (throttledDeviceId.containsKey(deviceId)) {
-                    throttledDeviceId.remove(deviceId);
-                  }
-                }
-              }
-            });
-    retryThread.start();
+    Set<String> users;
+    synchronized (bufferLock) {
+      users = new HashSet<>(idToBuffer.keySet());
+    }
+    for (String userId : users) {
+      Set<String> devices;
+      synchronized (bufferLock) {
+        devices = new HashSet<>(idToBuffer.get(userId).keySet());
+      }
+      for (String deviceId : devices) {
+        try {
+          retryThreadPool.submit(new RetryEventsOnLoop(userId, deviceId));
+        } catch (RejectedExecutionException e) {
```

Comment:
What's the possible cases for this exception? And if it is rejected, do we have any further handling?

Reply:
New tasks submitted in method execute(java.lang.Runnable) will be rejected when the Executor has been shut down, and also when the Executor uses finite bounds for both maximum threads and work queue capacity, and is saturated. For simple, two reason: 1. executor was shutdown. 2 task queue was full if there is a capacity limits. I think we don't need to do further handling since we don't shutdown the executor. New tasks is hard to be refused since we have retry buffer limits.